### PR TITLE
Implement Lightweight Postgres-backed Object Queue

### PR DESCRIPTION
**Pull request by jujaga**
# Description

This PR focuses on introducing the first part of the synchronization arc of work by implementing a lightweight postgres database based queue. This queue will enable us to track and federate work across multiple pod instances of COMS. Other changes include the anticipated database schema changes for synchronization and a few minor bugfixes and optimizations.

- 2fde130 Implement queueObject service
- a346cac Update DB models to synchronize with migration changes
- b0d1315 Extend checkSchema and knex to handle multiple schemas
- c1e944f Allow empty string tag values and fix notNullable constraint on tags
- 3b5629d Add synchronization and queue database migration
- 765ea00 Update S3 client defaults to no longer require explicit apiVersion

[SHOWCASE-3243](https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-3243)
[SHOWCASE-3264](https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-3264)

## Types of changes

Bug fix (non-breaking change which fixes an issue)
New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: app/src/db/models/tables/objectModel.js [hunk 1]
```diff
@@ -153,6 +153,9 @@ class ObjectModel extends Timestamps(Model) {
             });
         }
       },
+      findPath(query, value) {
```
Comment (TimCsaky) [inline app/src/db/models/tables/objectModel.js:156]: wonder if this should be named `filterPath` like other modifiers.

File: app/src/services/objectQueue.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      const response = await ObjectQueue.query(trx).insert({
+        bucketId: bucketId,
+        path: path,
+        full: full,
+        retries: retries,
+        createdBy: createdBy
+      }).onConflict().ignore();
+
+      if (!etrx) await trx.commit();
+      return Promise.resolve(!!response.id);
```
Comment (TimCsaky) [inline app/src/services/objectQueue.js:62]: so if there is a conflict (row exists already), this resolves to false., right?
that's intentional i assume.

**Review by TimCsaky — APPROVED** [review]

File: app/src/db/models/tables/objectModel.js (hunk 1 above)
Comment (jujaga) [reply app/src/db/models/tables/objectModel.js:156]: filterPath preserves the existing ilike searching behavior (ref Line 68) mostly used by the search endpoints; findPath is a bit more precise as it does an exact match instead.

File: app/src/services/objectQueue.js (hunk 2 above)
Comment (jujaga) [reply app/src/services/objectQueue.js:62]: Yep it's intentional.

File: app/src/db/migrations/20230609000000_009-sync-queue.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
+const stamps = require('../stamps');
+
+exports.up = function (knex) {
+  return Promise.resolve()
+    // Create audit schema and object_queue table
```
[the remark this answers is not in the document]
Comment (jujaga) [reply app/src/db/migrations/20230609000000_009-sync-queue.js]: Fixed

**Pull request merged**